### ISSUE TITLE
libclang: crash detected during reparsing

### ISSUE THREAD

**Issue by asardaes** [rstudio/rstudio#2268]
### System details

    RStudio Edition : Desktop
    RStudio Version : 1.1.423
    OS Version      : Windows 10
    R Version       : 3.4.3

### Steps to reproduce the problem

Edit any C++ file

### Describe the problem in detail

Making any changes to a C++ file triggers the error mentioned in the title, and after a few minutes RStudio crashes.

```r
> sessionInfo()
R version 3.4.3 (2017-11-30)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows >= 8 x64 (build 9200)

Matrix products: default

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] compiler_3.4.3 tools_3.4.3    yaml_2.1.16  
```

Last few lines of diagnostics report (edited to remove specifics from my computer):

```
24 Feb 2018 18:40:13 [rsession] ERROR system error 87 (The parameter is incorrect); 
OCCURRED AT: rstudio::core::Error rstudio::core::system::ChildProcess::run()
C:\Users\Administrator\rstudio\src\cpp\core\system\Win32ChildProcess.cpp:521; 
LOGGED FROM: void rstudio::session::modules::clang::{anonymous}::discoverSystemIncludePaths(std::vector<std::basic_string<char> >*)
C:\Users\Administrator\rstudio\src\cpp\session\modules\clang\CodeCompletion.cpp:235
24 Feb 2018 23:33:41 [rsession] ERROR Error re-parsing translation unit C:/Users/.../src/distmat-loops/distmat-loop.cpp; 
LOGGED FROM: rstudio::core::libclang::TranslationUnit rstudio::core::libclang::SourceIndex::getTranslationUnit(const string&, bool)
C:\Users\Administrator\rstudio\src\cpp\core\libclang\SourceIndex.cpp:223
```

**Comment by dfalty** [comment]
I can't reproduce this with 1.1.423 (Win 10) so I'm not sure if this is RStudio related. Others may be able to add input here.

**Comment by kevinushey** [comment]
FWIW I can usually reproduce the libclang crash by opening a new C++ file, saving it (note that C++ diagnostics only kick in on saved C++ files), and then attempting to type something like (try typing it rather than copy-pasting)

```cpp
#ifndef HEADER_GUARD
#define HEADER_GUARD

#endif
```

It seems like unterminated preprocessor definitions can lead to the libclang crash. Note that these crashes are somewhat expected; RStudio attempts to recover from libclang crashes since there's a somewhat high probability that unparseable C++ code could crash the libclang parser.

I'm not sure what could be leading to RStudio crashing altogether though. Maybe you're running RStudio in a path / folder that contains spaces or other non-ASCII characters? (Does your Windows username contain non-ASCII characters?)

**Comment by asardaes** [comment]
No, the path doesn't have any non-ASCII characters or even spaces for that matter. It is in my Dropbox, although I don't think that would make a difference.

This only happens in Windows, if I open the same project in Linux (same exact files, dual-boot system), I don't have any problems. I'll try reinstalling RStudio in Windows.

**Comment by asardaes** [comment]
I reinstalled RStudio and reset its state but I'm still getting the errors. I managed to trigger it quite reliably by opening any C++ file and adding `#include <memory>` *and* any valid `std::shared_ptr` to a function, e.g. `std::shared_ptr<int> test;`. The header alone doesn't trigger anything, and an invalid expression like `std::shared_ptr<> test;` doesn't trigger the error and shows a suggestion (too few template arguments).
